design/open/staging-compaction.md

```diff
@@ -0,0 +1,172 @@
+inefficient. Therefore, a Sensor deciding to compact will only do so if the 
+branch's `CompactionStartTime` is not within the last x<TBD> minutes using 
+`SetIf` to minimize collisions (although not avoiding them completely due to 
+clock skews). 
+
+### Compactor
+
+Upon deciding to compact, the Sensor will trigger the Compactor. The 
+Compactor will perform an operation very similar to Commit. Before starting 
+to compact, it will atomically:
+
+1. Push a new `StagingToken` to the branch.
+1. Add the old `StagingToken` to the `SealedTokens` list.
+1. Set the `CompactionStartTime` to the current time.
+
+The Compactor will now read the branch's `SealedTokens` in order and apply 
+them on either the `CompactedMetaRange` if it exists, or on the branch's 
+HEAD CommitID MetaRange to create a newMetaRange. The Compactor will then 
+atomically update the branch entity:
+
+1. Set the `CompactedMetaRangeID` to the new MetaRangeID.
+1. Remove the compacted `SealedTokens` from the branch.
+
+The Compactor should fail the operation if the sealed tokens have changed 
+since the compaction started (SetIf). Although the algorithm can tolerate 
+additional sealed tokens being added during the compaction, it's better to 
+avoid competing with concurrent commits. Commits have the same benefits as 
+compactions, but they are proactively triggered by the user (and might fail 
+if compaction succeeds).
+
+Consistency guarantees for the compaction process are derived from the 
+Commit operation consistency guarantees. In short, the compaction process 
+follows the same principles as the Commit operation. Writes to a branch 
+staging token succeeds only if it's the staging token after the write 
+occurred. Therefore, replacing the sealed token with an equivelant MetaRange 
+guarantees no successful writes are gone missing.
+
+### Branch Write Flow
+
+No change here. Writes are performed on the staging token.
+
+### Branch Read Flow
+
+Reading from a branch includes both a specific entry lookup and a listing 
+request on entries of a branch. The read flow is more complicated than what 
+we have today, and is different for compacted and uncompacted branches.
+
+#### Combined (Committed+ Uncommitted)
+
+There are 3 layers of reading from a branch:
+1. Read from the staging token.
+2. Read from the sealed tokens (in order).
+3. Depends on compaction:
+   1. If a CompactedMetaRangeID exists, read from the compacted MetaRange.
+   1. If a CompactedMetaRangeID doesn't exist, read from the CommitID.
+
+#### Committed
+
+Just like today, reads are performed on the Branch's CommitID.
+
+#### Uncommitted
+
+For operations such as `DiffUncommitted` or checking if the staging 
+area is empty, the read flow will be as follows:
+
+1. Read from the staging token.
+2. Read from the sealed tokens (in order).
+3. If a CompactedMetaRangeID exists, read the 2-way diff between the compacted
+   metarange and the CommitID's metarange.
```

Comment:
There still is a step I think I'm missing, what do I do with the data from the sealed (and staging) tokens and the data from the diff. There is some kind of "uncommitted merge" that we need to do.

Reply:
+1 on that

Reply:
This already exists today. Assume the following simple branch scenario today:
```
StagingToken: {A:{foo1:barA}}
SealedTokens: [{B:{foo1:barB,foo2:gooB}} , {C:{foo1:barC,foo2:gooC}}]
Commit: [{foo1:barCommitted,foo2:gooCommitted,,foo3:jooCommitted}]
```

Reading `foo1,foo2,foo3` from this branch yields `barA,gooB,jooCommitted`. For reading from the uncommitted area, you just skip the last layer with reading the commit.
With compaction the same algo holds, but you need to diff the CompactedMetaRange with the commit for the changes that are already manifested in the compaction